tests/test_util.py

```diff
@@ -365,21 +377,29 @@ def test_get_manifest_digests_missing(tmpdir, v1_digest, v2_digest):
     def request_callback(request):
         media_type = request.headers['Accept']
         media_type_prefix = media_type.split('+')[0]
-        # If requested schema version is not available, attempt to
-        # fallback to other version if possible to simulate how
-        # a docker registry behaves
-        if media_type.endswith('v2+json') and v2_digest:
-            digest = 'v2-digest'
-        elif media_type.endswith('v2+json') and v1_digest:
-            digest = 'not-used'
-            media_type_prefix = media_type_prefix.replace('v2', 'v1', 1)
-        elif media_type.endswith('v1+json') and v1_digest:
+
+        assert media_type.endswith('v2+json') or media_type.endswith('v1+json')
+
+        # Attempt to simulate how a docker registry behaves:
+        #  * If the stored digest is v1, return it
+        #  * If the stored digest is v2, and v2 is requested, return it
+        #  * If the stored digest is v2, and v1 is requested, try
+        #    to convert and return v1 or an error.
+        if digest_is_v1:
             digest = 'v1-digest'
-        elif media_type.endswith('v1+json') and v2_digest:
-            digest = 'not-used'
-            media_type_prefix = media_type_prefix.replace('v1', 'v2', 1)
+            media_type_prefix = media_type_prefix.replace('v2', 'v1', 1)
         else:
-            raise ValueError('Unexpected media type {}'.format(media_type))
+            if media_type.endswith('v2+json'):
+                digest = 'v2-digest'
+            else:
+                if not can_convert_v2_v1:
+                    # In some cases it's not possible to do the V2 => V1 conversion. For example,
+                    # if the V2 image doesn't have the legacy history information. This is the
+                    # response that the docker registry gives in that case
+                    headers = {}
+                    return (400, headers, '{"errors":[{"code":"MANIFEST_INVALID","message":"manifest invalid","detail":{}}]}')
```

Comment:
Hmm https://github.com/projectatomic/atomic-reactor/blob/master/CONTRIBUTING.md shouldn't recommend running pyflakes and pep8 only on the atomic_reactor subdirectory ;-) Will fix these and update the PR